JDK-8291878: NMT: Malloc limits

tstuefe opened the pull request:
(Second try, I withdrew the first patch https://github.com/openjdk/jdk/pull/9778 to do some other NMT improvements first)

This PR introduces malloc limits, similar to what the ancient `-XX:MallocMaxTestWords` did intend. `MallocMaxTestWords` is broken, but the solution proposed in this PR works fine since it is based on NMT. I plan to remove `-XX:MallocMaxTestWords`, but in a separate RFE since it requires fiddling with compiler tests.

### Why this is useful:

This allows us to generate a fatal error in the VM if the malloc amount - either globally or for a speicific NMT category - surpasses a given threshold.

We recently analyzed [JDK-8291919](https://bugs.openjdk.org/browse/JDK-8291919), a jdk11u-specific regression that caused compiler arena OOMs. A switch to limit compiler-related mallocs would have been very nice to cause a fatal error in the compiler thread, together with a replay file. I first tried `MallocMaxTestWords`, but that's broken since it does not de-account memory allocations.

In addition to customer scenarios like these, such a switch could be used to add sanity checks to compiler jtreg tests, adding malloc usage envelopes to tests. Maybe we could have caught [JDK-8291919](https://bugs.openjdk.org/browse/JDK-8291919) before shipment.

### How it works:

Patch introduces a new diagnostic switch `-XX:MallocLimit`. That switch can be used in two ways:

1 impose a total global limit to the size hotspot is allowed to malloc: 
```
-XX:MallocLimit=<size>
```
2 impose a limit to a selected NMT category, or to multiple NMT categories: 
```
-XX:MallocLimit=<category>:<size>[,<category>:<size>...]
```

If the switch is set, and the VM mallocs more in total (1) or for the given category (2), it will now stop with a fatal error. That way we can e.g. limit compiler arenas to a certain maximum in situations where the compiler runs amok, and get a compiler retry file. See here, with an artificial compiler bug introduced:

```
thomas@starfish$ ./images/jdk/bin/java  -XX:NativeMemoryTracking=summary -XX:+UnlockDiagnosticVMOptions -XX:MallocLimit=compiler:1g -jar /shared/projects/spring-petclinic/target/spring-petclinic-2.5.0-SNAP
SHOT.jar
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (mallocTracker.cpp:146), pid=519822, tid=519836
#  guarantee(false) failed: MallocLimit: category "Compiler" reached limit (size: 1073765608, limit: 1073741824) 
#
...
# An error report file with more information is saved as:
# /shared/projects/openjdk/jdk-jdk/output-release/hs_err_pid519822.log
#
# Compiler replay data is saved as:
# /shared/projects/openjdk/jdk-jdk/output-release/replay_pid519822.log
#
```

### Costs:

The feature requires that NMT checks the current number of allocated bytes via a given limit, on each malloc. 

- NMT disabled:
  - no costs
- NMT enabled but malloc limits are not used
  - we pay a very small memory overhead (138 bytes)
  - we don't pay performance
- NMT enabled and malloc limits are also enabled
  - if we have category-specific limits (e.g. `-XX:MallocLimit=compiler:1g`), the performance cost is so small it cannot be measured even in micro benchmarks. Which makes sense since we just compare two values that are likely to be cached by registers at the time of comparison.
  - Only if we have global limits (e.g. `-XX:MallocLimit=1g`) we pay a noticeable performance overhead. That is because on each malloc we need to add ~15 atomic counters to get the total malloc use (*). Noticeable means that with renaissance "philosophers" benchmark we see a performance drop of about 4%. This can be alleviated by future improvements of NMT that I have planned (only bottleneck for NMT improvement is the number of Reviewers willing to look at it).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291878](https://bugs.openjdk.org/browse/JDK-8291878): NMT: Malloc limits


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [7d937ff9](https://git.openjdk.org/jdk/pull/9891/files/7d937ff93ee6a03a003ccd39ab2d6bb1599dfd43)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9891/head:pull/9891` \
`$ git checkout pull/9891`

Update a local copy of the PR: \
`$ git checkout pull/9891` \
`$ git pull https://git.openjdk.org/jdk pull/9891/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9891`

View PR using the GUI difftool: \
`$ git pr show -t 9891`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9891.diff">https://git.openjdk.org/jdk/pull/9891.diff</a>

</details>

tstuefe commented:
/label hotspot-runtime

vnkozlov reviewed (approved):
Seems reasonable to me. But you need additional review from NMT experts.

tstuefe commented:
> Seems reasonable to me. 

Thank you, Vladimir. 

> But you need additional review from NMT experts.

Looks like I am the only one since Zhengyu left Redhat :(

@coleenp do you have someone with NMT knowledge left?

coleenp commented:
>@coleenp do you have someone with NMT knowledge left?

Not really - you've been doing all the work on NMT :)

tstuefe commented:
> > @coleenp do you have someone with NMT knowledge left?
> 
> Not really - you've been doing all the work on NMT :)

Haha :)

I'll ask Zhengyu, maybe he is still willing to take a look. Thanks!

shipilev commented on test/hotspot/jtreg/runtime/NMT/MallocLimitTest.java:
```diff
@@ -0,0 +1,251 @@
+        long small_memory_size = 1024*1024; // 1m
+        ProcessBuilder pb = processBuilderWithSetting("-XX:MallocLimit=" + small_memory_size);
+        OutputAnalyzer output = new OutputAnalyzer(pb.start());
+        output.reportDiagnosticSummary();
```
Do you really need `reportDiagnosticSummary()`? I thought it would be printed if `OutputAnalyzer` barfs on error.

shipilev reviewed (commented):
I think it is basically fine, but I have many stylistic comments.

tstuefe commented on src/hotspot/share/runtime/arguments.cpp:
```diff
@@ -4404,3 +4404,85 @@ bool Arguments::copy_expand_pid(const char* src, size_t srclen,
+  }
+}
+
+// Parse diagnostic NMT switch "MallocLimit" and return the found limits.
+// 1) If option is not given, it will set all limits to 0 (aka "no limit").
+// 2) If option is given in the global form (-XX:MallocLimit=<size>), it
+//    will return the size in *total_limit.
+// 3) If option is given in its per-NMT-category form (-XX:MallocLimit=<category>:<size>[,<category>:<size>]),
+//    it will return all found limits in the limits array.
+// 4) If option is malformed, it will exit the VM.
+// For (2) and (3), limits not affected by the switch will be set to 0.
```
Fixed.

tstuefe commented on src/hotspot/share/services/mallocTracker.cpp:
```diff
@@ -88,6 +94,56 @@ void MallocMemorySummary::initialize() {
+  Arguments::parse_malloc_limits(&_total_limit, _limits_per_category);
+
+  if (_total_limit > 0) {
+    log_info(nmt)("MallocLimit: total limit: " SIZE_FORMAT, _total_limit);
```
I did not know we have this. This is nice.

tstuefe commented:
Thanks @shipilev for the review!

I think I addressed all your issues. I re-ran manually x64 fastdebug, release and x86. The rest I leave up to GHAs and nightlies.

Cheers, Thomas

shipilev reviewed (approved):
This looks okay to me.

tstuefe commented:
Thank you Aleksey.

x86 errors unrelated.

/integrate
